p4-farm-nft/src/lib.rs

```diff
@@ -0,0 +1,2423 @@
+        let mut v = self.get_vault(&user);
+
+        // Expected cheddar for stake per one token
+        let expected = self.cheddar_rate;
+        assert_eq!(
+            expected, amount, 
+            "User need at least {} to stake one more token. Got {}",
+            self.cheddar_rate, amount
+        );
+
+        // update vault
+        v.total_cheddar_staked += amount;
+        // update total cheddar staked info
+        self.total_cheddar_stake += self.cheddar_rate;
+        log!("User stake {} Cheddar, which is required rate for stake 1 NFT token more", amount);
+        self.vaults.insert(&sender_id, &v);
+    }
+
+    /// Unstakes given token and transfers it back to the user.
+    /// If there is last staked token in vault - unstake and close the account
+    /// NOTE: account once closed must re-register to stake again.
+    /// Returns vector of staked tokens left (still staked) after the call.
+    /// Panics if the caller doesn't stake anything or if he doesn't have enough staked tokens.
+    /// Requires 1 yNEAR payment for wallet 2FA.
+    #[payable]
+    pub fn unstake(&mut self, nft_contract_id: &NftContractId, token_id: TokenId) -> Vec<TokenId> {
+        self.assert_is_active();
+        assert_one_yocto();
+        let user = env::predecessor_account_id();
+        self.internal_nft_unstake(&user, nft_contract_id, token_id)
+    }
+
+    /// Unstakes everything and close the account. Sends all farmed tokens using a ft_transfer
+    /// and all staked tokens back to the caller.
+    /// Panics if the caller doesn't stake anything.
+    /// Requires 1 yNEAR payment for wallet validation.
+    /// Max unstaking tokens per time limited - 4 tokens (greedy gas).
+    #[payable]
+    pub fn close(&mut self) {
+        self.assert_is_active();
+        assert_one_yocto();
+
+        let account_id = env::predecessor_account_id();
+        let mut vault = self.get_vault(&account_id);
+        self.ping_all(&mut vault);
+        log!("Closing {} account, farmed: {:?}", &account_id, vault.farmed);
+
+        // if user doesn't stake anything and has no rewards then we can make a shortcut
+        // and remove the account and return storage deposit.
+        if vault.is_empty() {
+            self.accounts_registered -= 1;
+            self.vaults.remove(&account_id);
+            Promise::new(account_id.clone()).transfer(STORAGE_COST);
+            return;
+        }
+
+        let units = min_stake(&vault.staked, &self.stake_rates);
+        self.staked_nft_units -= units;
+
+        // transfer all tokens to user
+        for nft_contract_id in 0..self.total_stake.len() {
+            let staked_tokens_ids = &vault.staked[nft_contract_id];
+            for token_id in 0..staked_tokens_ids.clone().len() {
+                self.transfer_staked_nft_token(
+                    account_id.clone(), 
+                    nft_contract_id, 
+                    staked_tokens_ids[token_id].clone()
+                );
```

Comment:
is it possible to send multiple tokens at once?

Reply:
Yes, it’s possible. But I’m thinking about gas also, we spent approximately 40 Tgas per token send for transfer and callbacks. Maybe we are need checks for amount on closed vault.

Reply:
Add assert to 5 tokens max for close possible